sympy/core/evalf.py

```diff
@@ -600,7 +600,7 @@ def evalf_add(v: 'Add', prec: int, options: OPT_DICT) -> TMP_RES:
         options['maxprec'] = min(oldmaxprec, 2*prec)
 
         terms = [evalf(arg, prec + 10, options) for arg in v.args]
-        n = terms.count(S.ComplexInfinity)
+        n = sum([term.count(S.ComplexInfinity) for term in terms])
```

Comment:
```suggestion
        n =filter(None, terms).count(S.ComplexInfinity)  # else None will be sympified during comparison
```

Reply:
`terms` is a `list` of `tuple` that potentially has `None`, so this would not filter the problematic `None`s and would restore the original problem (that during the comparison with `S.ComplexInfinity`, the `tuple` that contains `None` gets converted to `Tuple`)

Reply:
Note that I don't know if `terms` is always a `list` of `tuple` it just always was when I was printing it during debugging.

Reply:
The suggested change does not solve the underlying issue.

Reply:
It is not a problem that `terms` contains `None` it is a problem that `terms` contains `tuple` with `None`

Reply:
`n = [ term for term in terms if not isinstance(term, tuple)].count(S.ComlexInfinity)` could be used I think

Reply:
The problem is that `__eq__` should not create this Tuple with None in the first place. If sympify refused to sympify a tuple with unsympifiable arguments then everything would be fine:
https://github.com/sympy/sympy/blob/4489b4c7c51a65966b1a95f2af6f7e0c8c461794/sympy/core/expr.py#L148-L154

Reply:
> list of tuples
Sorry, @ThePauliPrinciple

Reply:
I think the confusion arised because my proposed solution doesn't solve the source of the problem as @oscarbenjamin indicates above.

Changing from `sympify` to `_sympify` in `Tuple` solves this issue I believe, but that should probably be deprecated properly. How do we solve this specific problem during the deprecation period?

Reply:
You could change `Tuple.__new__` to do
```python
try:
    args = tuple(_sympify(arg) for arg in args)
except SympifyError:
    # give a warning and then fall back to existing behaviour
    SymPyDeprecationWarning(...).warn()
    args = tuple(sympify(arg) for arg in args)

Reply:
I was hoping to do it in such a way that such warnings would only ever be shown if a user uses the library in a deprecated way and have the library always (at least for the available tests) avoid such situations.

Reply:
Oh, actually that's a good point. We don't want `a==b` to give a warning. There's probably a way to make this work but I also think that sometimes it's necessary to just make a change that isn't 100% compatible.

Reply:
A tuple doesn't need sympification to find out that it's not equal to zoo. We can catch this by reporting failure for any iterable that doesn't have a `_sympy` method (or whatever a 3rd party object must implement so `sympify` knows how to sympify it) since no iterable will equal an Expr.

Reply:
`Expr.__eq__`:
```python
    def __eq__(self, other):
        if not hasattr(other, '_sympy_') and iterable(other):
            # no expression is an iterable
            return False
        try:
            other = _sympify(other)
            if not isinstance(other, Expr):
                return False
...
```

Reply:
Well, here is the problem: `assert _sympify(tuple()) == Tuple()`  so any `tuple` gets converted to a `Tuple` (on `other = _sympify(other)` even if that causes a `Tuple` that should not exist e.g. `(None,)` gets converted to `Tuple(None)`.

Reply:
Changing `sympify` in `Tuple` to `_sympify` (because a `None` anywhere would raise a `SympifyError`) solves the issue, but SymPy generously abuses `Tuple` to apply `sympify` to multiple objects (along the lines of `x,y,z = Tuple(x, y, z)`). All those cases need to be fixed and then this can be solved with the aforementioned change in `Tuple`.

Reply:
The `iterable(other)` catches the tuple case and returns False. I tried this locally and it works fine. And if you ignore an iterable you can, with one test, ignore everything inside, `x == [i for i in range(1,10**6)]` will not need to sympify a million integers to indicate False ;-)

Reply:
I am posting this in several PRs hoping you will consider it and have a plan before going too far down the nonBasic erradication trail:  here is an example of storing Functions in an object: Subs(Derivative(f(f(x)), x), f, cos). How can this be handled if we don't allow BasicMeta as args?

Reply:
See https://github.com/sympy/sympy/pull/22539#issuecomment-985279336

Reply:
> 

But the "tuple case" `assert _sympify(tuple()) == Tuple()` is expected to return `True` as far as I understand

Reply:
I do agree with you on `Function`, I don't know how to solve that case, but I do believe that all other uses of non-`Basic` can be removed from the library. At which point figuring out how to make `Function` a `Basic` (or at the very least behave exactly like a `Basic`) might start to be worthwhile.